ipn/ipnserver/server.go

```diff
@@ -581,6 +583,76 @@ func (s *server) writeToClients(n ipn.Notify) {
 	}
 }
 
+// tryWindowsAppDataMigration attempts to copy the Windows state storage
+// directory from its old location to the new location. (Issue 2856)
+//
+// Tailscale 1.14 and before stored state under %LocalAppData%
+// (usually "C:\WINDOWS\system32\config\systemprofile\AppData\Local"
+// when tailscaled.exe is running as a non-user system service).
+// However it is frequently cleared for almost any reason: Windows
+// updates, System Restore, even various System Cleaner utilities.
+//
+// On failure, return the old location as a fallback where appropriate.
+func tryWindowsAppDataMigration(path string) string {
+	if path != paths.DefaultTailscaledStateFile() {
+		// If they're specifying a non-default path, just trust that they know
+		// what they are doing.
+		return path
+	}
+
+	oldFile := filepath.Join(os.Getenv("LocalAppData"), "Tailscale", "server-state.conf")
+	_, err := os.Stat(oldFile)
+	if os.IsNotExist(err) {
+		// Common case for a freshly installed system.
+		return path
+	}
+	if err != nil {
+		log.Printf("tryWindowsAppDataMigration failed; AppDataLocal: %v", err)
+		return path
+	}
+
+	newFile := filepath.Join(os.Getenv("ProgramData"), "Tailscale", "server-state.conf")
```

Comment:
do you even need this anymore filepath.Join and Getenv anymore? isn't this the same value as the `path` variable by definition now, after line 597's check?

Likewise, I'd move this up as the second return path, after "just trust they know what they're doing", like:


```go
	if path != paths.DefaultTailscaledStateFile() {
		// If they're specifying a non-default path, just trust that they know
		// what they are doing.
		return path
	}
        if _, err := os.Stat(path); err == nil {
              // Common case: the requested path already exists on disk, so it was either
              // a fresh install, or was already migrated.
              return path
       }
```

Reply:
Removed.
Could not make it a single line conditional because line 610 references err.